File: Programacion II/Practica 10/PrimerParcial-B POO/src/primerparcial/b/poo/Ministerio.java
package primerparcial.b.poo;
import java.util.*;
public class Ministerio {
    String nombre;
    String direccion;
    int nroEmpleados;
    String[][] empleados = new String[100][3];
    int[] edades = new int[100];
    double[] sueldos = new double[100];
    public Ministerio(String nombre, String direccion, String[][] empleados, int[] edades, double[] sueldos) {
        this.nombre = nombre;
        this.direccion = direccion;
        this.nroEmpleados = empleados.length;
        for (int i = 0; i < nroEmpleados; i++) {
            this.empleados[i] = empleados[i];
            this.edades[i] = edades[i];
            this.sueldos[i] = sueldos[i];
        }
    }
    public Ministerio(String nombre) {
        this.nombre = nombre;
        this.direccion = "";
        this.nroEmpleados = 0;
    }
    public void mostrarEmpleados() {
        for (int i = 0; i < nroEmpleados; i++) {
            System.out.println((i + 1) + ": " + String.join(" ", empleados[i]) +
                ", Edad: " + edades[i] + ", Sueldo: " + sueldos[i]);
        }
    }
    // (b) Eliminar empleados con edad X
    public void eliminarEmpleadoPorEdad(int edadX) {
        for (int i = 0; i < nroEmpleados; i++) {
            if (edades[i] == edadX) {
                for (int j = i; j < nroEmpleados - 1; j++) {
                    empleados[j] = empleados[j + 1];
                    edades[j] = edades[j + 1];
                    sueldos[j] = sueldos[j + 1];
                }
                nroEmpleados--;
                i--;
            }
        }
    }
    // (c) Transferir empleado desde otro objeto
    public void transferirEmpleadoDesde(Ministerio origen, String nombreEmpleado) {
        for (int i = 0; i < origen.nroEmpleados; i++) {
            String nombreCompleto = String.join(" ", origen.empleados[i]);
            if (nombreCompleto.equalsIgnoreCase(nombreEmpleado)) {
                this.empleados[this.nroEmpleados] = origen.empleados[i];
                this.edades[this.nroEmpleados] = origen.edades[i];
                this.sueldos[this.nroEmpleados] = origen.sueldos[i];
                this.nroEmpleados++;

                for (int j = i; j < origen.nroEmpleados - 1; j++) {
                    origen.empleados[j] = origen.empleados[j + 1];
                    origen.edades[j] = origen.edades[j + 1];
                    origen.sueldos[j] = origen.sueldos[j + 1];
                }
                origen.nroEmpleados--;
                break;
            }
        }
    }
    // (d) Mostrar empleado con menor edad o menor sueldo
    public void mostrarEmpleadoMenor(boolean porEdad) {
        if (nroEmpleados == 0) {
            System.out.println("No hay empleados.");
            return;
        }
        int minValor = porEdad ? edades[0] : (int) sueldos[0];
        for (int i = 1; i < nroEmpleados; i++) {
            int valor = porEdad ? edades[i] : (int) sueldos[i];
            if (valor < minValor) {
                minValor = valor;
            }
        }
        System.out.println("Empleados con menor " + (porEdad ? "edad" : "sueldo") + ":");
        for (int i = 0; i < nroEmpleados; i++) {
            int valor = porEdad ? edades[i] : (int) sueldos[i];
            if (valor == minValor) {
                System.out.println("Nombre: " + String.join(" ", empleados[i]));
                System.out.println("Edad: " + edades[i]);
                System.out.println("Sueldo: " + sueldos[i]);
                System.out.println("-----------------------");
            }
        }
    }
}

File: Programacion II/Practica 10/Ministerio.py
class Ministerio:
    def __init__(self, nombre, direccion="", empleados=None, edades=None, sueldos=None):
        self.nombre = nombre
        self.direccion = direccion
        self.empleados = empleados if empleados else []
        self.edades = edades if edades else []
        self.sueldos = sueldos if sueldos else []
    def mostrar_empleados(self):
        for i in range(len(self.empleados)):
            nombre_completo = " ".join(self.empleados[i])
            print(f"{i + 1}: {nombre_completo}, Edad: {self.edades[i]}, Sueldo: {self.sueldos[i]}")
    # b) Eliminar empleados con edad X
    def eliminar_empleado_por_edad(self, edad_x):
        i = 0
        while i < len(self.edades):
            if self.edades[i] == edad_x:
                self.empleados.pop(i)
                self.edades.pop(i)
                self.sueldos.pop(i)
            else:
                i += 1
    # c) Transferir empleado del segundo al primer objeto (operador simulado)
    def transferir_empleado_desde(self, origen, nombre_empleado):
        for i in range(len(origen.empleados)):
            if " ".join(origen.empleados[i]).lower() == nombre_empleado.lower():
                self.empleados.append(origen.empleados[i])
                self.edades.append(origen.edades[i])
                self.sueldos.append(origen.sueldos[i])

                origen.empleados.pop(i)
                origen.edades.pop(i)
                origen.sueldos.pop(i)
                break
    # d) Mostrar empleados con menor edad o menor sueldo
    def mostrar_empleado_menor(self, criterio="edad"):
      if not self.empleados:
          print("No hay empleados.")
          return
      if criterio == "edad":
          min_valor = min(self.edades)
      elif criterio == "sueldo":
          min_valor = min(self.sueldos)
      else:
          print("Criterio inválido.")
          return
      print(f"Empleados con menor {criterio}:")
      for i in range(len(self.empleados)):
          valor = self.edades[i] if criterio == "edad" else self.sueldos[i]
          if valor == min_valor:
              print(f"Nombre: {' '.join(self.empleados[i])}")
              print(f"Edad: {self.edades[i]}")
              print(f"Sueldo: {self.sueldos[i]}")
              print("------------------------")

# Datos de prueba
empleados = [
    ["Pedro", "Lucy", "Ana", "Saul"],
    ["Rojas", "Sosa", "Perez", "Arce"],
    ["Luna", "Rios", "Rojas" , "Calle"],
    ["Kion", "Alex", "Jauregui", "Coraite"]
]
edades = [35, 43, 26, 29]
sueldos = [2500, 3250, 2700, 2500]
empleados2 = [
    ["Juan", "Pepe", "Cruz", "Mamani"],
    ["Alejandro", "Esteban", "Aruquipa", "Cortez"]
]
edades2 = [35, 43, 26, 29]
sueldos2 = [2500, 3250, 2700, 2500]
# a) Instanciar 2 objetos de diferente forma
ministerio1 = Ministerio("Ministerio de Transporte", "Estación Central, Cementerio, 16 de Julio", empleados, edades, sueldos)
ministerio2 = Ministerio("Ministerio de Energía", "Zona Sur", empleados2, edades2, sueldos2)
# Mostrar empleados de ambos
print("== Ministerio 1 ==")
ministerio1.mostrar_empleados()
print("\n== Ministerio 2 ==")
ministerio2.mostrar_empleados()
# b) Eliminar por edad
print("\n== Ministerio 1 después de eliminar edad 26 ==")
ministerio1.eliminar_empleado_por_edad(26)
ministerio1.mostrar_empleados()
# c) Transferir a Saul del Ministerio 2 al 1
print("\n== Transferir 'Saul Arce Calle' a Ministerio 1 ==")
ministerio1.transferir_empleado_desde(ministerio2, "Juan Pepe Cruz Mamani")
print("Ministerio 1:")
ministerio1.mostrar_empleados()
print("Ministerio 2:")
ministerio2.mostrar_empleados()
# d) Mostrar menor edad y sueldo
print("\n== Menor edad en Ministerio 1 ==")
ministerio1.mostrar_empleado_menor("edad")
print("\n== Menor sueldo en Ministerio 1 ==")
ministerio1.mostrar_empleado_menor("sueldo")